apis/stacks/v1alpha1/stackdefinition.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StackDefinitionSpec defines the desired state of StackDefinition
+type StackDefinitionSpec struct {
+	AppMetadataSpec `json:",inline"`
+	CRDs            CRDList                 `json:"customresourcedefinitions,omitempty"`
+	Controller      ControllerSpec          `json:"controller,omitempty"`
+	Permissions     PermissionsSpec         `json:"permissions,omitempty"`
+	Behavior        StackDefinitionBehavior `json:"behavior,omitempty"`
+}
+
+// StackDefinitionSource is the stack image which this stack configuration is from.
+// In the future, other source types may be supported, such as a URL.
+type StackDefinitionSource struct {
+	// a container image id
+	Image string `json:"image,omitempty"`
+}
+
+// StackDefinitionBehavior specifies the behavior for the stack, if the stack has behaviors instead of a controller
+type StackDefinitionBehavior struct {
+	CRD    StackDefinitionBehaviorCRD       `json:"crd,omitempty"`
+	Engine StackResourceEngineConfiguration `json:"engine,omitempty"`
+	// Theoretically, source and engine could be specified at a per-hook level
+	// as well.
+	Source    StackDefinitionSource            `json:"source,omitempty"`
+	Reconcile StackDefinitionHookConfiguration `json:"reconcile,omitempty"`
+}
+
+// StackDefinitionBehaviorCRD represents the CRD which the stack's behavior controller will watch. When CRs of this
+// CRD kind appear and are modified in the cluster, the behavior will execute.
+type StackDefinitionBehaviorCRD struct {
+	APIVersion string `json:"apiVersion"`
+	Kind       string `json:"kind"`
+}
+
+// StackResourceEngineConfiguration represents a configuration for a resource engine, such as helm2 or kustomize.
```

Comment:
Well, the name I was originally going to use is `ResourceEngineConfiguration`, but I didn't want it to conflict with the `StackConfiguration` subfields. But maybe `ResourceEngineConfig` would be okay. I don't want to use variations of `Template` because it isn't necessarily templating. For example, I wouldn't really consider Kustomize a template-oriented approach.

Reply:
Yeah it's not. But then I think we need to find a new name and use it in other places, too. Were you thinking `ResourceEngine` as name of the concept?

Reply:
A `ResourceEngine` for `ResourcePacks`, I see where you are going with this.